concert/tests/integration/test_experiments.py

```diff
@@ -242,3 +264,35 @@ def test_attach_num_times(self):
         # Second time, cannot be called
         addon.detach()
         self.assertEqual(0, addon.attached_num_times)
+
+
+class TestExperimentStates(TestCase):
+    def tearDown(self):
+        shutil.rmtree(self.data_dir)
+
+    def setUp(self):
+        super(TestExperimentStates, self).setUp()
+        self.data_dir = tempfile.mkdtemp()
+        self.walker = DirectoryWalker(root=self.data_dir)
+
+    def test_experiment_state_normal(self):
+        exp = ExperimentSimple(self.walker)
+        self.assertEqual(exp.state, "standby")
+        exp_handle = exp.run()
+        sleep(0.1)
+        self.assertEqual(exp.state, "running")
+        exp_handle.join()
+        self.assertEqual(exp.state, "standby")
+        exp_handle = exp.run()
+        sleep(0.1)
+        self.assertEqual(exp.state, "running")
+        exp.abort().join()
+        self.assertEqual(exp.state, "aborted")
+        exp_handle.join()
+
+    def test_experiment_exception(self):
+        exp = ExperimentException(self.walker)
+        self.assertEqual(exp.state, "standby")
+        exp_handle = exp.run()
+        sleep(0.1)
```

Comment:
Why do you need to sleep and not simply wait for `exp_handle.join()`?

Reply:
join will block until the state is back to standby and the `_state_value` is not set fast enough. Without it you have a race condition if the state was set fast enough (that also why I don't like the `_state_value`)